### PR TITLE
[ao] Added fx model report per_channel detector

### PR DESCRIPTION
**Pull request by vspenubarthi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79243
* __->__ #79104

Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportFxDetector

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D37055549](https://our.internmc.facebook.com/intern/diff/D37055549)

### CODE REVIEW

File: torch/ao/quantization/fx/_model_report/_detector.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                else:
+                    raise ValueError("Should be either observer or fake quant")
+
+            per_channel_info["per_channel_status"][fqn] = {
+                "per_channel_supported": per_channel_supported,
+                "per_channel_used": per_channel_used,
+            }
+
+        # recurse on children
+        for child in module.children():
```
Comment (vkuzo) [inline torch/ao/quantization/fx/_model_report/_detector.py:113]: Instead of recursing on children, can we achieve the same result by using `module.named_modules` without recursion?

Comment (HDCharles) [reply torch/ao/quantization/fx/_model_report/_detector.py]: I think they both get only the direct children, though named_modules gets name, child rather than just child.

File: torch/ao/quantization/fx/_model_report/_detector.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+# Default map for representing supported per channel quantization modules for different backends
+DEFAULT_BACKEND_PER_CHANNEL_SUPPORTED_MODULES: Dict[str, Set[Any]] = {
+    "fbgemm": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, QatLinear, QatConvNd]),
+    "qnnpack": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, QatLinear, QatConvNd]),
+    "onednn": set([nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, QatLinear, QatConvNd]),
+}
+
+
+def _detect_per_channel(model: GraphModule) -> Tuple[str, Dict[str, Any]]:
```
Comment (HDCharles) [inline torch/ao/quantization/fx/_model_report/_detector.py:20]: nit: this can be a normal module, should still work since we aren't using any fx specific methods

**Review by HDCharles — APPROVED** [review]

File: torch/ao/quantization/fx/_model_report/_detector.py (hunk 1 above)
Comment (vkuzo) [reply torch/ao/quantization/fx/_model_report/_detector.py]: `named_modules` is a generator over all modules in the network, not just children of the current module. I'd recommend removing the recursion unless there is no way to do it without recursion.

Comment (HDCharles) [reply torch/ao/quantization/fx/_model_report/_detector.py]: oh ya that makes sense

File: test/quantization/fx/test_model_report_fx.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    LinearReLU(torch.nn.Linear(27, 27), torch.nn.ReLU()),
+    torch.nn.Conv2d(3, 3, 2, 1),
+)
+
+
+class TestModelReportFxDetector(QuantizationTestCase):
+
+    """Prepares and callibrate the model"""
+
+    def prepare_model_and_run_input(self, model, q_config_mapping, input):
```
Comment (HDCharles) [inline test/quantization/fx/test_model_report_fx.py:73]: this is a helper function rather than a test so prepend `_`

**Comment by vspenubarthi** [comment]
@vspenubarthi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37055549).
